### PR TITLE
feat: Add a User Profile page with an info card

### LINKED ISSUE #4283: Show user info in card on profile
We should show the user's name, photo, etc. in a card.

screenshot 2018-01-31 14 56 29

### PR DESCRIPTION
**Pull request by tofumatt**
Fix #4283

Looks like this:

### CODE REVIEW

**Comment by tofumatt** [comment]
This is ready for review now 🎉

File: src/amo/routes.js [hunk 1]
```diff
@@ -2,34 +2,34 @@ import config from 'config';
...
-import SearchPage from './components/SearchPage';
-import ServerError from './components/ErrorPage/ServerError';
-import Collection from './components/Collection';
-
 // If you add a new route here, check that the nginx rules maintained by ops
 // are in sync. See:
 // https://github.com/mozilla-services/puppet-config/tree/master/amo
 export default (
-  <div>
+  <React.Fragment>
```
Comment (tofumatt) [inline src/amo/routes.js]: Not wrapping our routes in a `<div>` seemed nice 😆

**Comment by tofumatt** [comment]
Will require #4284 once merged.

File: src/amo/components/UserProfile/index.js [hunk 2]
```diff
@@ -0,0 +1,176 @@
...
+
+    if (oldParams.username !== newParams.username) {
+      dispatch(fetchUserAccount({
+        errorHandlerId: errorHandler.id,
+        username: newParams.username,
+      }));
+    }
+  }
+
+  getUser(username: string) {
```
Comment (tofumatt) [inline src/amo/components/UserProfile/index.js]: Sorry, this needs to be moved to `mapStateToProps`, I was just doing it here lazily while hacking and forgot to fix it. I'll be getting it in a few.

Comment (kumar303) [reply src/amo/components/UserProfile/index.js]: ok, let me know when it's ready for review

**Comment by tofumatt** [comment]
Alrighty, change made and should be good.

File: src/amo/components/UserProfile/index.js [hunk 3]
```diff
@@ -0,0 +1,170 @@
...
+      i18n,
+      params,
+      user,
+    } = this.props;
+
+    if (errorHandler.hasError()) {
+      log.warn('Captured API Error:', errorHandler.capturedError);
+
+      if (errorHandler.capturedError.responseStatusCode === 404) {
+        return <NotFound errorCode={errorHandler.capturedError.code} />;
```
Comment (kumar303) [inline src/amo/components/UserProfile/index.js]: Ugh. This is annoying because it's custom logic that needs its own tests and maintenance. We will fix it in https://github.com/mozilla/addons-frontend/issues/2632

File: src/amo/components/UserProfile/index.js [hunk 4]
```diff
@@ -0,0 +1,170 @@
...
+      }
+    );
+
+    return (
+      <div className="UserProfile">
+        <Helmet>
+          <title>{userProfileTitle}</title>
+        </Helmet>
+
+        {errorHandler.renderErrorIfPresent()}
```
Comment (kumar303) [inline src/amo/components/UserProfile/index.js]: Since you're already checking for the error, the rendering logic should all happen in the same place to avoid confusion. It can look like this up above:

```js
let errorMessage;
if (errorHandler.hasError()) {
  if (errorHandler.capturedError.responseStatusCode === 404) {
    return <NotFound .../>;
  }

  errorMessage = errorHandler.renderError();
}
```
...
```js
return (
  <div className="UserProfile">
    {errorMessage}
  </div>
);
```

Otherwise it looks like you are forgetting to handle the error in the first if statement.

File: config/default.js [hunk 5]
```diff
@@ -288,4 +289,9 @@ module.exports = {
 
   // If true, show a link to the beta versions page on the listing page.
   betaVersions: true,
+
+  // Show and link to User Profiles using the front-end. These pages
+  // are still in development so this should only be enabled locally and
+  // on `-dev`.
+  enableUserProfile: false,
```
Comment (kumar303) [inline config/default.js]: Using config flags like this is a great way to ship half-working things. It helps a lot because we can iterate in smaller patches which are easier to review. Your patch here is already small, I just wanted to encourage this type of approach for future patches.

File: tests/unit/amo/components/TestUserProfile.js [hunk 6]
```diff
@@ -0,0 +1,315 @@
...
+    const { store } = dispatchSignInActions({
+      userProps: {
+        userId: 500,
+        username: 'tofumatt',
+      },
+    });
+    const root = renderUserProfile({ store });
+
+    expect(
+      root.find('.UserProfile-user-info').shallow().find('.UserProfile-name')
```
Comment (kumar303) [inline tests/unit/amo/components/TestUserProfile.js]: Doing a shallow render of `Card` isn't a good approach because it reaches into the implementation of `Card`. This might break if `Card` requires a context property in the future or if it its implementation changes in other ways. 

The `Card` component already has a well established interface that you are relying on: the `header` property. You can just reach in and render it like:

```js
const headerContent = root.find('.UserProfile-user-info').prop('header');
const header = shallow(<div>{headerContent}</div>);
expect(header.find('.UserProfile-name')).toHaveText('tofumatt');
```

This comment applies to other tests that use `root.find(...).shallow()`.

**Comment by tofumatt** [comment]
FYI this is ready.

File: src/amo/components/UserProfile/index.js [hunk 7]
```diff
@@ -0,0 +1,170 @@
...
+        <UserAvatar className="UserProfile-avatar" user={user} />
+
+        <h1 className="UserProfile-name">
+          {user ? user.displayName : <LoadingText />}
+        </h1>
+      </React.Fragment>
+    );
+    const userProfileTitle = i18n.sprintf(
+      i18n.gettext('User Profile for %(user)s'), {
+        user: user ? user.displayName : params.username,
```
Comment (kumar303) [inline src/amo/components/UserProfile/index.js]: Good idea on the fallback while loading

File: src/amo/components/UserProfile/index.js [hunk 8]
```diff
@@ -0,0 +1,170 @@
...
+            header={userProfileHeader}
+          >
+            <DefinitionList className="UserProfile-dl">
+              {user && user.homepage ? (
+                <Definition
+                  className="UserProfile-homepage"
+                  term={i18n.gettext('Homepage')}
+                >
+                  {/* TODO: Make a helper that strips https?:\/\/ from text */}
+                  <a href={user.homepage}>{user.homepage.replace(/^https?:\/\//, '')}</a>
```
Comment (kumar303) [inline src/amo/components/UserProfile/index.js]: Well, huh, I'm not sure how Flow lets you call `.replace()`. `user.homepage` can be `null` so I think you need protect against that.

File: src/amo/components/UserProfile/index.js [hunk 9]
```diff
@@ -0,0 +1,170 @@
...
+              >
+                {user ? user.num_addons_listed : <LoadingText />}
+              </Definition>
+              <Definition
+                className="UserProfile-rating-average"
+                term={i18n.gettext('Average rating of developer’s add-ons')}
+              >
+                {user ? (
+                  <Rating
+                    rating={parseFloat(user.average_addon_rating)}
```
Comment (kumar303) [inline src/amo/components/UserProfile/index.js]: Why did you need `parseFloat()`? `Rating` already [calls it](https://github.com/mozilla/addons-frontend/blob/master/src/ui/components/Rating/index.js#L98).

File: src/amo/components/UserProfile/index.js [hunk 10]
```diff
@@ -0,0 +1,170 @@
...
+                  />
+                ) : <LoadingText />}
+              </Definition>
+            </DefinitionList>
+
+            {user && user.biography && user.biography.length ? (
+              <div
+                className="UserProfile-biography"
+                // eslint-disable-next-line react/no-danger
+                dangerouslySetInnerHTML={sanitizeUserHTML(user.biography)}
```
Comment (kumar303) [inline src/amo/components/UserProfile/index.js]: The old site supported a bunch of tags. We should probably start by doing the same:

screenshot 2018-03-14 11 59 49

File: src/amo/components/UserProfile/index.js [hunk 11]
```diff
@@ -0,0 +1,170 @@
...
+          </Card>
+        </div>
+      </div>
+    );
+  }
+}
+
+export function mapStateToProps(
+  state: {| users: UsersStateType |},
+  ownProps: {| params: UserProfileParams |},
```
[the remark this answers is not in the document]
Comment (tofumatt) [reply src/amo/components/UserProfile/index.js]: Blarg, oops! Thanks.

File: src/amo/components/UserProfile/index.js [hunk 12]
```diff
@@ -0,0 +1,170 @@
...
+) {
+  return {
+    user: state.users.byID[state.users.byUsername[ownProps.params.username]],
```
[the remark this answers is not in the document]
Comment (tofumatt) [reply src/amo/components/UserProfile/index.js]: Good call.

File: src/amo/components/UserProfile/index.js [hunk 13]
```diff
@@ -0,0 +1,170 @@
...
+
+
+type UserProfileParams = {| username: string |};
+
+type Props = {|
+  dispatch: DispatchFunc,
+  errorHandler: ErrorHandlerType,
+  i18n: I18nType,
+  params: UserProfileParams,
+  user: UserType,
```
[the remark this answers is not in the document]
Comment (tofumatt) [reply src/amo/components/UserProfile/index.js]: Oops, yes, you're totally right. Thanks!

File: src/amo/components/UserProfile/index.js (hunk 4 above)
Comment (tofumatt) [reply src/amo/components/UserProfile/index.js]: I saw that in other components and I thought it seemed more verbose... but fair point about it not looking like an error is checked until then otherwise. I'll change it 👍

File: src/amo/components/UserProfile/index.js (hunk 8 above)
Comment (tofumatt) [reply src/amo/components/UserProfile/index.js]: This code is only executed if `user && user.homepage` is truthy, so that's unlikely, but I will tidy this up regardless.

File: src/amo/components/UserProfile/index.js (hunk 9 above)
Comment (tofumatt) [reply src/amo/components/UserProfile/index.js]: Ohhhhhhh, I think it's because of how things are called. `Rating` calls `toFixed(1)` first, but I think the data returned by the API caused an issue with `.toFixed(1)`. Maybe `Rating` needs to be fixed.

File: src/amo/components/UserProfile/index.js (hunk 10 above)
Comment (tofumatt) [reply src/amo/components/UserProfile/index.js]: That's exactly what `sanitizeUserHTML` does 😄:

https://github.com/mozilla/addons-frontend/blob/master/src/core/utils/index.js#L90

File: src/amo/components/UserProfile/index.js (hunk 9 above)
Comment (tofumatt) [reply src/amo/components/UserProfile/index.js]: Ah, and it's because of https://github.com/mozilla/addons-server/issues/7822

I can either add the `parseFloat` to the user reducer or here until https://github.com/mozilla/addons-server/issues/7822 is fixed, or keep the fix to `Rating` I added.

**Comment by tofumatt** [comment]
Thanks for the review, this is good for another look.

Let me know what you think about the best way to work around https://github.com/mozilla/addons-server/issues/7822

File: src/amo/components/UserProfile/index.js (hunk 10 above)
Comment (kumar303) [reply src/amo/components/UserProfile/index.js]: oh duh, I read it as `sanitizeHTML()` for some reason

Comment (tofumatt) [reply src/amo/components/UserProfile/index.js]: Hah, well that’s easy to mix up 😅

File: src/amo/components/UserProfile/styles.scss [hunk 14]
```diff
@@ -0,0 +1,49 @@
...
+  @include respond-to(medium) {
+    font-size: $font-size-xl;
+  }
+}
+
+.UserProfile-rating-average .Rating {
+  @include margin-start(-3px);
+
+  justify-content: start;
+  margin: 0;
```
Comment (kumar303) [inline src/amo/components/UserProfile/styles.scss]: Hmm, doesn't this overwrite `margin-start(-3px)`?

File: src/amo/routes.js [hunk 15]
```diff
@@ -39,6 +39,8 @@ export default (
       <IndexRoute component={Home} />
       <Route path="addon/:slug/" component={Addon} />
       <Route path="addon/:addonSlug/reviews/" component={AddonReviewList} />
+      {config.get('enableUserProfile') &&
```
Comment (kumar303) [inline src/amo/routes.js]: Is this necessary? I don't think the URL is accessible on the frontend until we allow it from nginx, right?

File: src/core/utils/index.js [hunk 16]
```diff
@@ -192,6 +193,13 @@ export function getErrorComponent(status) {
   }
 }
 
+export function removeProtocolFromURL(urlWithProtocol) {
+  invariant(urlWithProtocol, 'url is required');
```
Comment (kumar303) [inline src/core/utils/index.js]: `url` should be `urlWithProtocol`

File: tests/unit/amo/components/TestUserProfile.js [hunk 17]
```diff
@@ -0,0 +1,318 @@
...
+  });
+
+  it('renders the user avatar', () => {
+    const root = renderUserProfile();
+
+    const headerContent = root.find('.UserProfile-user-info').prop('header');
+    const header = shallow(<div>{headerContent}</div>);
+
+    expect(header.find(UserAvatar)).toHaveProp(
+      'user', getCurrentUser(root.instance().props.store.getState().users));
```
Comment (kumar303) [inline tests/unit/amo/components/TestUserProfile.js]: It feels a little fragile to rely on `.props.store.getState()`. How about doing it with more setup?

```diff
diff --git a/tests/unit/amo/components/TestUserProfile.js b/tests/unit/amo/components/TestUserProfile.js
index 2291383e5..ec603f585 100644
--- a/tests/unit/amo/components/TestUserProfile.js
+++ b/tests/unit/amo/components/TestUserProfile.js
@@ -94,13 +94,17 @@ describe(__filename, () => {
   });
 
   it('renders the user avatar', () => {
-    const root = renderUserProfile();
+    const username = 'tofumatt';
+    const { state, store } = dispatchSignInActions({
+      userProps: { username },
+    });
+    const root = renderUserProfile({ store, params: { username } });
 
     const headerContent = root.find('.UserProfile-user-info').prop('header');
     const header = shallow(<div>{headerContent}</div>);
 
     expect(header.find(UserAvatar)).toHaveProp(
-      'user', getCurrentUser(root.instance().props.store.getState().users));
+      'user', getCurrentUser(state.users));
   });
 
   it('still passes user prop to avatar while loading', () => {
```

If that setup becomes too repetitive you could add a helper to sync up `params` and `user`.

I also find it helpful to see explicit setup of data for tests that are testing that same data.

File: tests/unit/amo/components/TestUserProfile.js [hunk 18]
```diff
@@ -0,0 +1,318 @@
...
+    const header = shallow(<div>{headerContent}</div>);
+
+    expect(header.find(UserAvatar)).toHaveProp('user', undefined);
+  });
+
+  it("renders the user's name", () => {
+    const root = renderUserProfile();
+
+    const headerContent = root.find('.UserProfile-user-info').prop('header');
+    const header = shallow(<div>{headerContent}</div>);
```
Comment (kumar303) [inline tests/unit/amo/components/TestUserProfile.js]: These two lines are repeated a lot. You could move them to a helper like:

```js
const header = getHeaderComponent(root);
```

File: tests/unit/amo/components/TestUserProfile.js [hunk 19]
```diff
@@ -0,0 +1,318 @@
...
+
+    expect(header.find('.UserProfile-name').find(LoadingText))
+      .toHaveLength(1);
+  });
+
+  it("renders the user's username if no display name exists", () => {
+    const { store } = dispatchSignInActions({
+      userProps: {
+        display_name: null,
+        userId: 500,
```
Comment (kumar303) [inline tests/unit/amo/components/TestUserProfile.js]: This doesn't seem relevant to the test

File: tests/unit/amo/components/TestUserProfile.js [hunk 20]
```diff
@@ -0,0 +1,318 @@
...
+    const root = renderUserProfile({ params: { username: 'not-loaded' } });
+
+    expect(root.find(ReportUserAbuse)).toHaveLength(1);
+  });
+
+  it('renders a not found page if the API request is a 404', () => {
+    const { store } = dispatchSignInActions();
+    const errorHandler = new ErrorHandler({
+      id: 'some-id',
+      capturedError: { responseStatusCode: 404 },
```
Comment (kumar303) [inline tests/unit/amo/components/TestUserProfile.js]: This looks pretty fragile. [Here](https://github.com/mozilla/addons-frontend/blob/ec3a4dd837ac92516283f3db53e3c21aaec6efc8/tests/unit/amo/components/TestCollection.js#L485-L493) is a better way to do it.

File: src/ui/components/Rating/index.js [hunk 21]
```diff
@@ -95,7 +95,7 @@ export class RatingBase extends React.Component {
     let description;
     if (rating) {
       description = i18n.sprintf(i18n.gettext('Rated %(rating)s out of 5'),
-        { rating: i18n.formatNumber(parseFloat(rating.toFixed(1))) });
+        { rating: i18n.formatNumber(parseFloat(rating).toFixed(1)) });
```
Comment (kumar303) [inline src/ui/components/Rating/index.js]: This seems like the right thing to do but can you add a test for it? I didn't see one.

File: src/amo/components/UserProfile/styles.scss (hunk 14 above)
Comment (tofumatt) [reply src/amo/components/UserProfile/styles.scss]: It doesn't, *but* @tsl143 used `margin()` elsewhere and I should really just use that.

File: src/amo/routes.js (hunk 15 above)
Comment (tofumatt) [reply src/amo/routes.js]: I guess not, but it seemed more complete/proper. I can remove it, but I just figured leaving it here was a bit of code-as-documentation that this is pref'd. I can just leave a comment though, fair enough.

File: src/core/utils/index.js (hunk 16 above)
Comment (tofumatt) [reply src/core/utils/index.js]: I guess... really it's just saying the argument should be truthy; as it's not a named argument I thought "url" was more helpful. But I guess `urlWithProtocol` will be easier to grep.

File: tests/unit/amo/components/TestUserProfile.js (hunk 17 above)
Comment (tofumatt) [reply tests/unit/amo/components/TestUserProfile.js]: Fair enough. It's a bit fragile but I found it less work.

On the second point I totally agree though, so yeah I'll refactor this. It's handy to see all the props set in the test. 👍

File: tests/unit/amo/components/TestUserProfile.js (hunk 18 above)
Comment (tofumatt) [reply tests/unit/amo/components/TestUserProfile.js]: Hah, I was thinking that when I refactored it and I was lazy. Thanks for calling me out on that, will fix 👍

File: tests/unit/amo/components/TestUserProfile.js (hunk 19 above)
Comment (tofumatt) [reply tests/unit/amo/components/TestUserProfile.js]: It's not, shoot, I remember why I did it and that's no longer relevant. Thanks! 🔥

File: tests/unit/amo/components/TestUserProfile.js (hunk 20 above)
Comment (tofumatt) [reply tests/unit/amo/components/TestUserProfile.js]: I feel like I tried that and something broke, but I was probably doing it wrong™.

File: src/ui/components/Rating/index.js (hunk 21 above)
Comment (tofumatt) [reply src/ui/components/Rating/index.js]: Will do! @diox helpfully patched the API response but we still feed integers to it in some of the tests and also it just makes sense to do. I'll add it. 👍

File: tests/unit/amo/components/TestUserProfile.js (hunk 20 above)
Comment (tofumatt) [reply tests/unit/amo/components/TestUserProfile.js]: It worked, I think I must've just forgot to pass `store.dispatch` last time 😓

**Comment by tofumatt** [comment]
Thanks for the review; changes made, good for another look.

File: src/amo/components/UserProfile/styles.scss [hunk 22]
```diff
@@ -0,0 +1,48 @@
...
+  font-size: $font-size-l;
+  margin: 0;
+
+  @include respond-to(medium) {
+    font-size: $font-size-xl;
+  }
+}
+
+.UserProfile-rating-average .Rating {
+  @include margin(0, 0, 0, -3px);
```
Comment (kumar303) [inline src/amo/components/UserProfile/styles.scss]: Seeing a `-3px` negative margin in a review is hard because I can't tell if the final margin follows our standard spacing. Maybe `Rating` needs a new class that can be toggled on via configuration? I see that search results need to adjust it too. No need to address it here but maybe file as a code quality issue.

File: tests/unit/amo/components/TestUserProfile.js [hunk 23]
```diff
@@ -0,0 +1,321 @@
...
+  it("renders the user's biography", () => {
+    const { store } = dispatchSignInActions({
+      userProps: {
+        biography: 'Not even vegan!',
+        username: 'tofumatt',
+      },
+    });
+    const root = renderUserProfile({ store });
+
+    expect(root.find('.UserProfile-biography').render().html())
```
Comment (kumar303) [inline tests/unit/amo/components/TestUserProfile.js]: You can remove the `.render()` part. You only need `.html()`.

File: tests/unit/amo/components/TestUserProfile.js [hunk 24]
```diff
@@ -0,0 +1,321 @@
...
+    const dispatchSpy = sinon.spy(store, 'dispatch');
+
+    const root = renderUserProfile({
+      params: { username: 'black-panther' },
+      store,
+    });
+
+    dispatchSpy.reset();
+
+    sinon.assert.notCalled(dispatchSpy);
```
Comment (kumar303) [inline tests/unit/amo/components/TestUserProfile.js]: This is not necessary after calling `.reset()`

File: tests/unit/amo/components/TestUserProfile.js [hunk 25]
```diff
@@ -0,0 +1,321 @@
...
```
Comment (kumar303) [inline tests/unit/amo/components/TestUserProfile.js]: same here

File: tests/unit/amo/components/TestUserProfile.js [hunk 26]
```diff
@@ -0,0 +1,321 @@
...
+    const root = renderUserProfile({ store });
+
+    expect(root.find('.UserProfile-biography')).toHaveLength(0);
+  });
+
+  it('omits an empty biography', () => {
+    const { store } = dispatchSignInActions({
+      userProps: {
+        biography: '',
+        username: 'tofumatt',
```
Comment (kumar303) [inline tests/unit/amo/components/TestUserProfile.js]: The `username` is repeated everywhere to sync up with `params`. You could make a simple change to DRY it up:

```js
const { store } = dispatchSignInActions({
  userProps: defaultUserProps({
    biography: '',
  }),
});
```
Or you could wrap `dispatchSignInActions()` in a helper.

File: src/amo/components/UserProfile/styles.scss (hunk 22 above)
Comment (tofumatt) [reply src/amo/components/UserProfile/styles.scss]: Good call. I think it might be that in most places it indents itself too far over; in reality I think it needs a class for the original place it was used as I think the CSS was too coupled to the parent component originally. I'll file the follow-up issues, cheers 👍

File: tests/unit/amo/components/TestUserProfile.js (hunk 23 above)
Comment (tofumatt) [reply tests/unit/amo/components/TestUserProfile.js]: Whoops, right on.

File: tests/unit/amo/components/TestUserProfile.js (hunk 24 above)
Comment (tofumatt) [reply tests/unit/amo/components/TestUserProfile.js]: Indeed, I don't need to test sinon in our tests 😅

File: tests/unit/amo/components/TestUserProfile.js (hunk 26 above)
Comment (tofumatt) [reply tests/unit/amo/components/TestUserProfile.js]: Nice, will do!

File: tests/unit/amo/components/TestUserProfile.js (hunk 23 above)
Comment (tofumatt) [reply tests/unit/amo/components/TestUserProfile.js]: Actually, no, the `.html()` will return an object. I'll try a different matcher.

Comment (tofumatt) [reply tests/unit/amo/components/TestUserProfile.js]: I think it's because it's doing dangerouslySetInnerHTML...

File: src/amo/components/UserProfile/styles.scss (hunk 22 above)
Comment (tofumatt) [reply src/amo/components/UserProfile/styles.scss]: Filed #4573.
